libs/checkpoint-vercel-kv/src/index.ts

```diff
@@ -0,0 +1,389 @@
+        for _, key in ipairs(keys) do
+          table.insert(result, key)
+        end
+      until cursor == '0'
+      return result
+    `;
+
+    const pendingWriteKeys: string[] = await this.kv.eval(
+      pendingWLua,
+      [],
+      [pendingWPrefix]
+    );
+
+    if (pendingWriteKeys.length) {
+      const serializedWrites: (KVPendingWrite | null)[] = await this.kv.mget(
+        ...pendingWriteKeys
+      );
+      pendingWrites = await Promise.all(
+        serializedWrites
+          .filter((write): write is KVPendingWrite => write !== null)
+          .map(async (serializedWrite) => {
+            const [sType, sDumpedValue] = this.serde.dumpsTyped(
+              serializedWrite.value
+            );
+            const unserializedValue = await this.serde.loadsTyped(
+              sType,
+              sDumpedValue
+            );
+            return [
+              serializedWrite.task_id,
+              serializedWrite.channel,
+              unserializedValue,
+            ] as CheckpointPendingWrite;
+          })
+      );
+    }
+    return {
+      checkpoint,
+      metadata,
+      pendingWrites,
+      config: {
+        configurable: {
+          thread_id,
+          checkpoint_ns,
+          checkpoint_id: checkpoint.id,
+        },
+      },
+      parentConfig: row.parent_checkpoint_id
+        ? {
+            configurable: {
+              thread_id,
+              checkpoint_ns,
+              checkpoint_id: row.parent_checkpoint_id,
+            },
+          }
+        : undefined,
+    };
+  }
+
+  async *list(
+    config: RunnableConfig,
+    options?: CheckpointListOptions
+  ): AsyncGenerator<CheckpointTuple> {
+    const { thread_id, checkpoint_ns = "" } = config.configurable ?? {};
+    const { limit, before, filter } = options ?? {};
+
+    // LUA script to get keys excluding those starting with "last"
+    const luaScript = `
+      local prefix = ARGV[1]
+      local cursor = '0'
+      local result = {}
+      repeat
+        local scanResult = redis.call('SCAN', cursor, 'MATCH', prefix .. '*', 'COUNT', 1000)
+        cursor = scanResult[1]
+        local keys = scanResult[2]
+        for _, key in ipairs(keys) do
+          if key:sub(-5) ~= ':last' then
+            table.insert(result, key)
+          end
+        end
+      until cursor == '0'
+      return result
+    `;
+
+    // Execute the LUA script with the prefix as an argument
+    const prefix = `${thread_id}:${checkpoint_ns}`;
+    const keys: string[] = await this.kv.eval(luaScript, [], [prefix]);
+
+    // Filter keys based on the before parameter
+    const filteredKeys = keys.filter((key: string) => {
+      const [, , checkpoint_id] = key.split(":");
+      return !before || checkpoint_id < before?.configurable?.checkpoint_id;
+    });
+
+    const sortedKeys = filteredKeys.sort((a: string, b: string) =>
+      b.localeCompare(a)
+    );
+
+    const rows: (KVRow | null)[] = await this.kv.mget(...sortedKeys);
```

Comment:
This should get only up to `limit` keys at a time and loop?

Reply:
Yes. Takes the matching keys, but to support metadata filtering, we its doing so when it gets the value in filtering and count in the loop / generator. I'd be better perhaps to change the LUA script to an HSCAN and filter the metadata there.